src/tui/terminal.rs

```diff
@@ -0,0 +1,115 @@
+use crossterm::cursor;
+use crossterm::event;
+use crossterm::terminal;
+use miette::IntoDiagnostic;
+use miette::WrapErr;
+use ratatui::prelude::CrosstermBackend;
+use ratatui::prelude::Terminal;
+use std::io::Stdout;
+use std::ops::Deref;
+use std::ops::DerefMut;
+use std::panic;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+
+pub struct TerminalGuard {
+    terminal: Terminal<CrosstermBackend<Stdout>>,
+}
+
+impl Deref for TerminalGuard {
+    type Target = Terminal<CrosstermBackend<Stdout>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.terminal
+    }
+}
+
+impl DerefMut for TerminalGuard {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.terminal
+    }
+}
+
+impl Drop for TerminalGuard {
+    fn drop(&mut self) {
+        let result = exit().wrap_err("Failed to exit terminal during drop");
+        // Ignore the `Result` if we're already panicking; aborting is undesirable
+        if !std::thread::panicking() {
+            result.unwrap();
+        }
+    }
+}
+
+static INSIDE: AtomicBool = AtomicBool::new(false);
```

Comment:
Tracking whether we're "inside" or not (raw mode enabled, cursor hidden, events being captured, etc etc) so we can skip double entering or double exiting, in particular the implicit exits with `Drop` and panics.

This is preferable to removing the panic hook we installed, because we can't be sure an additional panic hook "layer" hasn't been installed (e.g. the `miette` panic hook in `main`). Safer to turn ours into a no-op post-exit and leave the panic hook undisturbed.

TODO: Skip double entering